### PR TITLE
calldata encoding

### PR DESCRIPTION
**Pull request by CalabashSquash**
Mostly using [Aave v3](https://docs.aave.com/developers/getting-started/l2-optimization) as reference.

# Changes
- Create `L2Encoder.sol` to provide an easy-to-use suite of `pure`/`view` functions for encoding function params into `bytesX` types.
- Create `libraries/CalldataLogic.sol`, a library for decoding function parameters from `bytesX` types to their "real" types.
- Move `PoolSwapLibrary.sol` into the new `libraries/` directory.
- Change `PoolCommitter::commit`s function signature to `commit(bytes32 args)`, and now use `CalldataLogic` to decode into the 4 parameters.
- Lots of test files changed
- Create `PoolKeeper::performUpkeepMultiplePoolsPacked`
- Create `L2Encoder::encodeAddressArray`
- Changed the parameters of `multiPaidClaimSinglePoolCommitter` and `multiPaidClaimMultiplePoolCommitters` to take `bytes` args to allow the compacting of address arrays

### CODE REVIEW

File: test/L2Encoder/encodeCommitParams.spec.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                amount,
+                commitType,
+                fromAggregateBalance,
+                payForClaim
+            )
+            expect(result).to.equal(expectedResult)
+        })
+    })
+    context("Standard input", async () => {
+        context("Case 1", async () => {
```
Comment (sporejack) [inline test/L2Encoder/encodeCommitParams.spec.ts]: A comment explaining what cases 1 and 2 are would be helpful.

File: contracts/implementation/L2Encoder.sol [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     * @param payForClaim True if user wants to pay for the commit to be claimed
+     * @return compact representation of commit parameters
+     */
+    function encodeCommitParams(
+        uint256 amount,
+        IPoolCommitter.CommitType commitType,
+        bool fromAggregateBalance,
+        bool payForClaim
+    ) external pure returns (bytes32) {
+        uint128 shortenedAmount = amount.toUint128();
```
Comment (sporejack) [inline contracts/implementation/L2Encoder.sol:45]: This [reverts](https://docs.openzeppelin.com/contracts/2.x/api/utils#SafeCast-toUint128-uint256-) on overflow -- is this intended?

Comment (CalabashSquash) [reply contracts/implementation/L2Encoder.sol:45]: Yep that's intended

File: test/L2Encoder/encodeCommitParams.spec.ts (hunk 1 above)
Comment (CalabashSquash) [reply test/L2Encoder/encodeCommitParams.spec.ts]: Addressed

**Review by sporejack — APPROVED** [review]

File: test/L2Encoder/encodeCommitParams.spec.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            const fromAggregateBalance = false
+            const payForClaim = false
+            let expectedResult = ethers.utils.solidityPack(
+                ["bool", "bool", "uint8", "uint128"],
+                [payForClaim, fromAggregateBalance, commitType, amount]
+            )
+            // Pad with 0s
+            expectedResult =
+                "0x" + "0".repeat(paddingRequired) + expectedResult.slice(2)
+            console.log(expectedResult.length)
```
Comment (CalabashSquash) [inline test/L2Encoder/encodeCommitParams.spec.ts]:
```suggestion
```

File: test/L2Encoder/encodeCommitParams.spec.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                    let expectedResult = ethers.utils.solidityPack(
+                        ["bool", "bool", "uint8", "uint128"],
+                        [payForClaim, fromAggregateBalance, commitType, amount]
+                    )
+                    // Pad with 0s
+                    expectedResult =
+                        "0x" +
+                        "0".repeat(paddingRequired) +
+                        expectedResult.slice(2)
+                    console.log(expectedResult.length)
```
Comment (CalabashSquash) [inline test/L2Encoder/encodeCommitParams.spec.ts]:
```suggestion
```

File: test/L2Encoder/encodeCommitParams.spec.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (CalabashSquash) [inline test/L2Encoder/encodeCommitParams.spec.ts]:
```suggestion
```

**Review by 0xhrithik — APPROVED** [review]

**Pull request merged**